smi/adaptor/controller.go

```diff
@@ -0,0 +1,340 @@
+					log.Infof("skipping %s sp as ignore annotation is present", sp.Name)
+					return nil
+				}
+
+				sp.Spec.DstOverrides = nil
+				_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(namespace).Update(context.Background(), sp, metav1.UpdateOptions{})
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+		return err
+	}
+
+	// Check if the Service Profile is already Present
+	sp, err := c.spLister.ServiceProfiles(ts.Namespace).Get(fmt.Sprintf("%s.%s.svc.%s", ts.Spec.Service, ts.Namespace, c.clusterDomain))
+	if err != nil {
+		// Create a Service Profile resource as it does not exist
+		_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Create(context.Background(), c.toServiceProfile(ts), metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Infof("Service Profile already present")
+		if ignoreAnnotationPresent(sp) {
+			log.Infof("skipping %s sp as ignore annotation is present", sp.Name)
+			return nil
+		}
+
+		// Check if SP Matches the TS, and update if it not
+		spFromTs := c.toServiceProfile(ts)
+		if !equal(spFromTs, sp) {
+			log.Infof("serviceprofile/%s does not match trafficscplit/%s", sp.Name, ts.Name)
+			updateDstOverrides(sp, ts, c.clusterDomain)
+			log.Infof("Updated Service Profile as it's not equivalent to the relevant TS")
+			_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Update(context.Background(), sp, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func ignoreAnnotationPresent(sp *serviceprofile.ServiceProfile) bool {
+	_, ok := sp.Annotations[ignoreServiceProfileAnnotation]
+	return ok
+}
+
+func equal(spA *serviceprofile.ServiceProfile, spB *serviceprofile.ServiceProfile) bool {
+	if spA.Name != spB.Name {
+		return false
+	}
+
+	if spA.Namespace != spB.Namespace {
+		return false
+	}
+
+	if len(spA.Spec.DstOverrides) != len(spB.Spec.DstOverrides) {
+		return false
+	}
+
+	for index, dstA := range spA.Spec.DstOverrides {
+		dstB := spB.Spec.DstOverrides[index]
+
+		if dstA.Authority != dstB.Authority {
+			return false
+		}
+
+		if !dstA.Weight.Equal(dstB.Weight) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// enqueue takes a Ts or SP resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than Ts or SP.
+func (c *SMIController) enqueue(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		log.Error(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+// updateDstOverrides updates the dstOverrides of the given serviceprofile
+// to match that of the trafficsplit
+func updateDstOverrides(sp *serviceprofile.ServiceProfile, ts *trafficsplit.TrafficSplit, clusterDomain string) {
+	sp.Annotations[createdOrUpdatedBySMIAdaptorFor] = ts.Name
+	sp.Spec.DstOverrides = []*serviceprofile.WeightedDst{}
```

Comment:
In my understanding here, we are overriding `DstOverrides` that already exist for a `serviceprofile` and replacing them 
with `DstOverrides` found in `trafficsplits`. Is there ever a case where a `serviceprofile` may already have overrides we want to preserve?

Reply:
Yep, Users can always explicitly set them and those are overridden by the adaptor when a TS is applied on the same service, unless the `skip` annotation is added by the user.

So, Users can always use that annotation to prevent the `smi-adaptor` from changing it